bzt/modules/passfail.py

```diff
@@ -162,24 +160,32 @@ def __repr__(self):
     def process_criteria_logic(self, tstmp, get_value):
         value = self.agg_logic(tstmp, get_value)
         state = self.condition(value, self.threshold)
-        if state:
-            if self.window_logic == 'within':
-                self.counting = self.window
-                self.is_triggered = True
+
+        if self.window_logic == 'for':
+            if state:
+                self.counting += 1
+                if self.counting >= self.window:
+                    if not self.is_triggered:
+                        logging.warning("%s", self)
+                    self.is_triggered = True
             else:
-                self._count(tstmp)
+                self.counting = 0
         else:
-            self.counting = 0
-            if not self.is_triggered:
-                self.started = None
+            self.counting += 1
+            if self.window_logic == 'within' and state:
+                self.is_triggered = True
+            elif self.window_logic == 'over' and state and self.counting >= self.window:
+                self.counting = self.window
+                self.is_triggered = True
+            elif not state:
+                self.counting = self.window
```

Comment:
Yes, I realized after your comments that stuff around `counting` goes wrong. I'm gonna rework it a bit now.